### DISCUSSION TITLE
Is it possible to use hibernate multitenancy with smallrye-reactive-messaging?

### DISCUSSION THREAD

**Discussion by n-kremer** [quarkusio/quarkus#22637] [Q&A] [answered]
Hi,

[hibernate-orm-multi-tenancy-quickstart](https://github.com/quarkusio/quarkus-quickstarts/tree/main/hibernate-orm-multi-tenancy-quickstart) shows how to implement multitenancy when using REST requests. While this works perfectly well, I could not find any information if this is also possible with smallrye-reactive-messaging-kafka. I tried to implement an example. but could not get it working.

Use Case:
We will receive information about the tenant in every Kafka message. Based on this information, the correct schema has to be selected.

What I tried to far:
I created [a small test application](https://github.com/n-kremer/hibernate-orm-multi-tenancy-with-kafka) to see if this is working.
The app contains a CustomTenantResolver that randomly selects one of two schemas.
When sending REST requests I see that the different schemas are used if I repeat the request several times.
In order to test if this is also working with Kafka, the app selects values from the database every time a new message is received and prints the first one out. I expected to see the same behavior as in the REST endpoint, so randomly selecting a schema on every message. However, the schema is selected only once and every subsequent message uses the same schema.

Can I modify my code so that the CustomTenantResolver gets executed every time a new message is received?
If yes, is there a context that would work for both REST request and Kafka messages?

[a comment by an automated account is not shown]

**Reply by cescoffier** [reply]
I don't believe it's possible. The QuickStart uses the incoming request and so the request scope to configure the tenant. This is not something you can do with Kafka as there is no request scope, so you cannot implement a tenant resolver taking the 'current' kafka message as parameter.

Now, I don't know if there are other way to configure the tenant than a request scoped bean.

**Reply by n-kremer** [reply]
Could I manually get a requestContext using Arc in this case, or would this potentially lead to problems?

Example:
```java
 @Incoming("prices")
  @Blocking
  public void receivePrice(Double price) {
    var cont = Arc.container().requestContext();
    cont.activate();
    requestScopedTenantService.setTenant("Tenant from message");
    Fruit.listAll();
    cont.destroy();
    cont.deactivate();
  }
```

**Reply by Ladicek** [reply]
Classes with `@Incoming` / `@Outgoing` methods are CDI beans, aren't they? So annotating an `@Incoming` / `@Outgoing` method with `@ActivateRequestContext` could be enough.

**Reply by n-kremer** [reply]
Unfortunately the approach with `@ActivateRequestContext` does not seem to be working. The `CustomTenantResolver` is still only called once in the beginning.

**Reply by Ladicek** [reply]
Sorry I was too brief. I only meant that `@ActivateRequestContext` should be enough to make request context active in the method, no need to do it programmatically.

You'd still have to set the tenant identifier or whatever (I don't know anything about Hibernate multitenancy).

**Reply by n-kremer** [reply]
There seems to be a difference between programmatically activating the context and using  `@ActivateRequestContext`.

The code from my previous comment in which I programmatically activated the context works as expected. On every incoming message a new instance of the `requestScopedTenantService` is created and afterwards the `CustomerTenantResolver` is called when `Fruit.listAll()` is executed.

When I replace the programmatical activation of the context with `@ActivateRequestContext`,  the behavior is different.  It does not seem like a request context is activated every time the method is called (e.g. only one instance of `requestScopedTenantService` is created when `receivePrice` is called for the first time).

```java
  @Incoming("prices")
  @ActivateRequestContext
  @Blocking
  public void receivePrice(Double price) {
    requestScopedTenantService.setTenant("Tenant from message");
    Fruit.listAll();
  }

**Reply by Ladicek** [reply]
That's strange. Maybe `@ActivateRequestContext` is called _before_ the method is offloaded to another thread (I don't really know if it works like that or not), that would explain it.

In that case, what you have is sensible to me. If you don't want to use ArC-specific API, you can inject a `RequestContextController` and use _that_ to activate/deactivate the request context. But that's just cosmetics.

**Reply by n-kremer** [reply]
Thanks for your help :)

I'm fine with using the ArC-specific API, just wanted to be sure if that's a valid way to solve this problem.
